### PR TITLE
implement `PartialEq<str>` for `Bound<'py, PyString>`

### PR DESCRIPTION
**Pull request by davidhewitt**
Idea spurred by #4020 and also what I've just been doing in #4196.

This PR implements comparison between Python `str` and Rust `&str` by using:
- The new `PyUnicode_EqualToUTF8AndSize` on Python 3.13,
- fallback to converting to a Rust `&str` and doing equality that way on older versions.

`PyUnicode_EqualToUTF8AndSize` can never fail, so similarly the fallback implementations just return `false` on error getting UTF8 data out.

### CODE REVIEW

**Review by LilyFoote — APPROVED** [review]

**Comment by davidhewitt** [comment]
Thanks for the approval @LilyFoote!

I just realised: there's a bit edge case where: what about when the `Bound<'py, PyString>` contains a subclass of `str`? That subclass can override inequality, and might not compare equal with strings (although that would be very weird).

What should we do then? One option is to make these implementations return false for subclasses (e.g. do an `is_exact_instance_of::<PyString>()` check).

In `PyUnicode_EqualToUTF8` there is no subclass checking - https://github.com/python/cpython/blob/42351c3b9a357ec67135b30ed41f59e6f306ac52/Objects/unicodeobject.c#L10789

I'm inclined to proceed with this as-is, but I would also be glad to hear folks' opinions as there is a slight possibility this is a footgun which leads to surprises in the subclass edge case.

**Comment by LilyFoote** [comment]
I think subclassing built-in types like `str` is fairly well known to have edge-cases and oddities like this. Maybe all we need here is a note in the docs.

**Comment by davidhewitt** [comment]
> I think subclassing built-in types like `str` is fairly well known to have edge-cases and oddities like this. Maybe all we need here is a note in the docs.

True! I think to that point, it's very likely that users would write `bound_str.to_str()? == "rust string"` which is just as wrong for subclasses as what we implement here. So documenting gotcha on the trait implementations as you suggest and moving on might be fine.

Any other opinions before we proceed?

File: src/types/string.rs [hunk 1]
```diff
@@ -762,4 +844,26 @@ mod tests {
...
+        Python::with_gil(|py| {
+            let s = "hello, world";
+            let py_string = PyString::new_bound(py, s);
+
+            assert_eq!(py_string, "hello, world");
+
+            assert_eq!(py_string, s);
+            assert_eq!(&py_string, s);
+            assert_eq!(s, py_string);
+            assert_eq!(s, &py_string);
```
Comment (Icxolu) [inline src/types/string.rs:859]: Do we want to also add the variants with `*s` to also test the implementations targeting `str` and not `&str`?

**Review by Icxolu — APPROVED** [review]
I agree here. I think it is reasonable to expect `PyString` to do normal string comparison and having the `PartialEq` impls makes a lot of code much more ergonomic and readable. Documenting this behavior seems sufficient.

File: src/types/string.rs (hunk 1 above)
Comment (davidhewitt) [reply src/types/string.rs:914]: Good catch, done 👍

**Comment by davidhewitt** [comment]
Thanks both! I've added documentation and will proceed to merge 👍

**Pull request merged**